File: 2020/day01/main.ml
let puzzle1 =
    let f = open_in "input" in
    let target = 2020 in
    let seen = Array.make target false in
    try
        while true; do
            let line = input_line f in
            let a = int_of_string line in
            let b = target - a in
            if 0 <= b && b < target && seen.(b) then begin
                Printf.printf "%d\n" (a * b);
                raise Exit
            end;
            seen.(a) <- true;
        done;
        close_in f
    with
    | Exit -> ()
    | End_of_file -> Printf.printf "Not found\n"
    | e ->
        close_in_noerr f;
        raise e
;;

let puzzle2 =
    let f = open_in "input" in
    let target = 2020 in
    let seen = Array.make target false in
    try
        while true; do
            let line = input_line f in
            let a = int_of_string line in
            for b = 0 to 2019; do
                let c = target - a - b in
                if seen.(b) && 0 <= c && c < target && seen.(c) then begin
                    Printf.printf "%d\n" (a * b * c);
                    raise Exit
                end;
            done;
            seen.(a) <- true;
        done;
        close_in f
    with
    | Exit -> ()
    | End_of_file -> Printf.printf "Not found\n"
    | e ->
        close_in_noerr f;
        raise e
;;

let main =
    puzzle1;
    puzzle2
;;

main

File: 2020/day01/main.py
from typing import Iterator, Set

def read_input() -> Iterator[int]:
    with open('input') as f:
        for line in f:
            yield int(line)

def puzzle1() -> None:
    seen: Set[int] = set()
    target = 2020
    for a in read_input():
        b = target - a
        if b in seen:
            print(a * b)
            return
        seen.add(a)
    assert False


def puzzle2() -> None:
    seen: Set[int] = set()
    target = 2020
    for a in read_input():
        for b in seen:
            c = target - a - b
            if c in seen:
                print(a * b * c)
                return
        seen.add(a)
    assert False


def main() -> None:
    puzzle1()
    puzzle2()


if __name__ == '__main__':
    main()
